### PR TITLE
feat: support container network mode

### LINKED ISSUE #1286: Watchtower Fails Recreating Containers that Share a Network Stack
Hello! I have a VPN container and torrent container that share a network stack (torrent container has --network container:PIA). I'm noticing that whenever something happens to the "parent" VPN container that the torrent container errors out. This can come in a few different ways:

1. Both are re-created

In this case, it seems that the torrent container is created before the VPN container:

```
time="2022-05-11T02:43:54Z" level=info msg="Found new linuxserver/qbittorrent:latest image (e3ce01e9d9f9)"
time="2022-05-11T02:44:17Z" level=info msg="Found new linuxserver/sonarr:latest image (e4571e1de8bb)"
time="2022-05-11T02:44:33Z" level=info msg="Found new linuxserver/jackett:latest image (9209598da4bc)"
time="2022-05-11T02:44:39Z" level=info msg="Found new qmcgaw/gluetun:latest image (dc68aaf68f41)"
time="2022-05-11T02:45:04Z" level=info msg="Found new linuxserver/unifi-controller:latest image (b6f250fead08)"
time="2022-05-11T02:45:08Z" level=info msg="Found new guacamole/guacamole:latest image (772b60108cca)"
time="2022-05-11T02:45:15Z" level=info msg="Found new guacamole/guacd:latest image (4969201c0757)"
time="2022-05-11T02:45:17Z" level=info msg="Found new itzg/minecraft-server:latest image (f7c85977cfb8)"
time="2022-05-11T02:45:20Z" level=info msg="Stopping /Guacd (11395aded9fc) with SIGTERM"
time="2022-05-11T02:45:30Z" level=info msg="Stopping /Guac (4401e75cc7cf) with SIGTERM"
time="2022-05-11T02:45:35Z" level=info msg="Stopping /Unifi (68e49fc241d3) with SIGTERM"
time="2022-05-11T02:45:47Z" level=info msg="Stopping /PIA (7f374bb5714f) with SIGTERM"
time="2022-05-11T02:45:49Z" level=info msg="Stopping /Jackett (dc8a5ad9dde7) with SIGTERM"
time="2022-05-11T02:45:53Z" level=info msg="Stopping /Sonarr (340cec5a4766) with SIGTERM"
time="2022-05-11T02:45:58Z" level=info msg="Stopping /Nginx (ec102dfb297d) with SIGTERM"
time="2022-05-11T02:46:08Z" level=info msg="Stopping /qBittorrent (9324abf693e5) with SIGTERM"
time="2022-05-11T02:46:17Z" level=info msg="Creating /qBittorrent"
time="2022-05-11T02:46:17Z" level=error msg="Error response from daemon: No such container: 7f374bb5714f1b081a6da007fea491d4d8eb586ee6541be406b9dbb666bcabef"
time="2022-05-11T02:46:17Z" level=info msg="Creating /Nginx"
time="2022-05-11T02:46:18Z" level=info msg="Creating /Sonarr"
time="2022-05-11T02:46:19Z" level=info msg="Creating /Jackett"
time="2022-05-11T02:46:21Z" level=info msg="Creating /PIA"
time="2022-05-11T02:46:22Z" level=info msg="Creating /Unifi"
time="2022-05-11T02:46:24Z" level=info msg="Creating /Guac"
time="2022-05-11T02:46:25Z" level=info msg="Creating /Guacd"
time="2022-05-11T02:46:27Z" level=info msg="Creating /Minecraft"
time="2022-05-11T02:46:27Z" level=info msg="Session done" Failed=1 Scanned=26 Updated=7 notify=no
```

2. Parent container is updated without the child container.

In this case the containers are recreated but the child container looses the connection to the parent container (presumably because the container hash changes).

I understand that a similar issue used to happen with the --link command but that was resolved back in March. This seems to be the same issue just with the new command (--network container:<> replaced --link).

### PR DESCRIPTION
**Pull request by schizo99**
This PR solves the issue when using `network_mode: service:xxxxx` in a service.
It closes #1286 

~For this to work not only does the label `com.centurylinklabs.watchtower.depends-on` be set to the parent container. It also require a new label `com.centurylinklabs.watchtower.network-service` to be set.~

It solves the issue by updating the HostConfig with the ID of the new parent container.

**Update**: Since `NetworkMode: "container:*"` would mean that the container would not work without being re-created if the network supplying container was updated, this should no longer be necessary. Instead it is now treated as an implicit link.

### CODE REVIEW

**Comment by schizo99** [comment]
I think I have found a better way to solve this.

Will update PR later on.

**Comment by schizo99** [comment]
Maybe did a mistake by force pushing?

In any case I think this new approach is nicer and does not require any additional labels.

File: pkg/container/client.go [hunk 1]
```diff
@@ -157,6 +157,15 @@ func (client dockerClient) GetContainer(containerID t.ContainerID) (Container, e
 		return Container{}, err
 	}
 
+	containerNetworkMode := strings.Split(string(containerInfo.HostConfig.NetworkMode), ":")
+	if len(containerNetworkMode) == 2 {
```
Comment (piksel) [inline pkg/container/client.go]:
```suggestion
	if len(containerNetworkMode) == 2 && containerNetworkMode[0] == "container" {
```
if this contains some other prefix that we don't know about, we should probably leave it alone

File: pkg/container/client.go [hunk 2]
```diff
@@ -157,6 +157,15 @@ func (client dockerClient) GetContainer(containerID t.ContainerID) (Container, e
...
+		parentContainer, err := client.api.ContainerInspect(bg, containerNetworkMode[1])
+		if err != nil {
+			log.Debug("Unable to fetch parentContainer.")
+		}
+		containerInfo.HostConfig.NetworkMode = container.NetworkMode(fmt.Sprintf("container:%s", parentContainer.Name))
```
Comment (piksel) [inline pkg/container/client.go]:
```suggestion
		} else {
			containerInfo.HostConfig.NetworkMode = container.NetworkMode(fmt.Sprintf("container:%s", parentContainer.Name))
		}
```

If we don't get a parentContainer we shouldn't be updating the `NetworkMode`. In fact, the container should probably be skipped with an error if this happens, but at least this way will just fall back to the old (current) behaviour of watchtower.

**Review by piksel — CHANGES_REQUESTED** [review]
This seems like it *should* work. The only thing I am not thrilled about is that the API for getting the container info also alters the returned info. The way it's used right now, this shouldn't be an issue, but it's hidden/unexpected, which could lead to bugs further on.

Of course, the "right" way would be to do this when preparing the new configuration before starting the container. Except that at that time, the old container should already have been removed.

**Comment by schizo99** [comment]
You're right. I'm not 100% sold on this but it's the most efficient way.
It might, as you say, cause issues should the Docker API change from allowing name or id in the network_mode.

The other option would be my first approach which would require an additional label to be added besides the depends_on label.

**Review by piksel — APPROVED** [review]
It still needs tests, but otherwise I am happy with this. Changing the `container:ContainerID` to `container:ContainerName` is in a way more correct, even if it's not what the raw inspect contained.

**Comment by schizo99** [comment]
Was intending to add tests.

However, as soon as I update the `container_watchtower.json` and change network_mode to `container:xxxxx` the server throws this error.

<details>
<summary>Go test thrown errors</summary>

```
(cloud) {15:37}~/fab/watchtower/pkg/container:networkmode-service ✗ ➭ go test -v
=== RUN   TestContainer
Running Suite: Container Suite
==============================
Random Seed: 1666618638
Will run 44 of 44 specs

•••••WARN[0000] Failed to retrieve container image info: request returned Internal Server Error for API route and version http://127.0.0.1:33371/v1.41/images/sha256:4dbc5f9c07028a985e14d1393e849ea07f68804c4293050d5a641b138db72daa/json, check if the server supports the requested API version 

------------------------------
• Failure [0.003 seconds]
the client
/home/aahp/fab/watchtower/pkg/container/client_test.go:23
  when listing containers
  /home/aahp/fab/watchtower/pkg/container/client_test.go:72
    when no filter is provided
    /home/aahp/fab/watchtower/pkg/container/client_test.go:73
      should return all available containers [It]
      /home/aahp/fab/watchtower/pkg/container/client_test.go:74

      Path mismatch
      Expected
          <string>: /v1.41/containers/xxxxxxxxxxxxxxxxxxxx/json
      to have suffix
          <string>: /images/sha256:4dbc5f9c07028a985e14d1393e849ea07f68804c4293050d5a641b138db72daa/json

      /home/aahp/go/pkg/mod/github.com/onsi/gomega@v1.21.1/ghttp/handlers.go:51
------------------------------
WARN[0000] Failed to retrieve container image info: request returned Internal Server Error for API route and version http://127.0.0.1:38345/v1.41/images/sha256:4dbc5f9c07028a985e14d1393e849ea07f68804c4293050d5a641b138db72daa/json, check if the server supports the requested API version 
• Failure [0.001 seconds]
the client
/home/aahp/fab/watchtower/pkg/container/client_test.go:23
  when listing containers
  /home/aahp/fab/watchtower/pkg/container/client_test.go:72
    when a filter matching nothing
    /home/aahp/fab/watchtower/pkg/container/client_test.go:86
      should return an empty array [It]
      /home/aahp/fab/watchtower/pkg/container/client_test.go:87

      Path mismatch
      Expected
          <string>: /v1.41/containers/xxxxxxxxxxxxxxxxxxxx/json
      to have suffix
          <string>: /images/sha256:4dbc5f9c07028a985e14d1393e849ea07f68804c4293050d5a641b138db72daa/json

      /home/aahp/go/pkg/mod/github.com/onsi/gomega@v1.21.1/ghttp/handlers.go:51
------------------------------
WARN[0000] Failed to retrieve container image info: request returned Internal Server Error for API route and version http://127.0.0.1:35389/v1.41/images/sha256:4dbc5f9c07028a985e14d1393e849ea07f68804c4293050d5a641b138db72daa/json, check if the server supports the requested API version 
• Failure [0.002 seconds]
the client
/home/aahp/fab/watchtower/pkg/container/client_test.go:23
  when listing containers
  /home/aahp/fab/watchtower/pkg/container/client_test.go:72
    when a watchtower filter is provided
    /home/aahp/fab/watchtower/pkg/container/client_test.go:100
      should return only the watchtower container [It]
      /home/aahp/fab/watchtower/pkg/container/client_test.go:101

      Path mismatch
      Expected
          <string>: /v1.41/containers/xxxxxxxxxxxxxxxxxxxx/json
      to have suffix
          <string>: /images/sha256:4dbc5f9c07028a985e14d1393e849ea07f68804c4293050d5a641b138db72daa/json

      /home/aahp/go/pkg/mod/github.com/onsi/gomega@v1.21.1/ghttp/handlers.go:51
------------------------------
WARN[0000] Failed to retrieve container image info: request returned Internal Server Error for API route and version http://127.0.0.1:42385/v1.41/images/sha256:4dbc5f9c07028a985e14d1393e849ea07f68804c4293050d5a641b138db72daa/json, check if the server supports the requested API version 
• Failure [0.002 seconds]
the client
/home/aahp/fab/watchtower/pkg/container/client_test.go:23
  when listing containers
  /home/aahp/fab/watchtower/pkg/container/client_test.go:72
    when include stopped is enabled
    /home/aahp/fab/watchtower/pkg/container/client_test.go:113
      should return both stopped and running containers [It]
      /home/aahp/fab/watchtower/pkg/container/client_test.go:114

      Path mismatch
      Expected
          <string>: /v1.41/containers/xxxxxxxxxxxxxxxxxxxx/json
      to have suffix
          <string>: /images/sha256:4dbc5f9c07028a985e14d1393e849ea07f68804c4293050d5a641b138db72daa/json

      /home/aahp/go/pkg/mod/github.com/onsi/gomega@v1.21.1/ghttp/handlers.go:51
------------------------------
WARN[0000] Failed to retrieve container image info: request returned Internal Server Error for API route and version http://127.0.0.1:45507/v1.41/images/sha256:4dbc5f9c07028a985e14d1393e849ea07f68804c4293050d5a641b138db72daa/json, check if the server supports the requested API version 
• Failure [0.001 seconds]
the client
/home/aahp/fab/watchtower/pkg/container/client_test.go:23
  when listing containers
  /home/aahp/fab/watchtower/pkg/container/client_test.go:72
    when include restarting is enabled
    /home/aahp/fab/watchtower/pkg/container/client_test.go:126
      should return both restarting and running containers [It]
      /home/aahp/fab/watchtower/pkg/container/client_test.go:127

      Path mismatch
      Expected
          <string>: /v1.41/containers/xxxxxxxxxxxxxxxxxxxx/json
      to have suffix
          <string>: /images/sha256:4dbc5f9c07028a985e14d1393e849ea07f68804c4293050d5a641b138db72daa/json

      /home/aahp/go/pkg/mod/github.com/onsi/gomega@v1.21.1/ghttp/handlers.go:51
------------------------------
WARN[0000] Failed to retrieve container image info: request returned Internal Server Error for API route and version http://127.0.0.1:39927/v1.41/images/sha256:4dbc5f9c07028a985e14d1393e849ea07f68804c4293050d5a641b138db72daa/json, check if the server supports the requested API version 
• Failure [0.001 seconds]
the client
/home/aahp/fab/watchtower/pkg/container/client_test.go:23
  when listing containers
  /home/aahp/fab/watchtower/pkg/container/client_test.go:72
    when include restarting is disabled
    /home/aahp/fab/watchtower/pkg/container/client_test.go:139
      should not return restarting containers [It]
      /home/aahp/fab/watchtower/pkg/container/client_test.go:140

      Path mismatch
      Expected
          <string>: /v1.41/containers/xxxxxxxxxxxxxxxxxxxx/json
      to have suffix
          <string>: /images/sha256:4dbc5f9c07028a985e14d1393e849ea07f68804c4293050d5a641b138db72daa/json

      /home/aahp/go/pkg/mod/github.com/onsi/gomega@v1.21.1/ghttp/handlers.go:51
------------------------------
• Failure [1.002 seconds]
the client
/home/aahp/fab/watchtower/pkg/container/client_test.go:23
  ExecuteCommand
  /home/aahp/fab/watchtower/pkg/container/client_test.go:153
    when logging
    /home/aahp/fab/watchtower/pkg/container/client_test.go:154
      should include container id field [It]
      /home/aahp/fab/watchtower/pkg/container/client_test.go:155

      Timed out after 1.001s.
      Got stuck at:
          ERRO[0000] Failed to extract command exec logs: cannot connect to the Docker daemon. Is 'docker daemon' running on this host?: dial http: unknown network http  containerID=ex-cont-id
          
      Waiting for:
          containerID="?ex-cont-id"?

      /home/aahp/fab/watchtower/pkg/container/client_test.go:218
------------------------------
••••••••••••••••••••••••••••••••

Summarizing 7 Failures:

[Fail] the client when listing containers when no filter is provided [It] should return all available containers 
/home/aahp/go/pkg/mod/github.com/onsi/gomega@v1.21.1/ghttp/handlers.go:51

[Fail] the client when listing containers when a filter matching nothing [It] should return an empty array 
/home/aahp/go/pkg/mod/github.com/onsi/gomega@v1.21.1/ghttp/handlers.go:51

[Fail] the client when listing containers when a watchtower filter is provided [It] should return only the watchtower container 
/home/aahp/go/pkg/mod/github.com/onsi/gomega@v1.21.1/ghttp/handlers.go:51

[Fail] the client when listing containers when include stopped is enabled [It] should return both stopped and running containers 
/home/aahp/go/pkg/mod/github.com/onsi/gomega@v1.21.1/ghttp/handlers.go:51

[Fail] the client when listing containers when include restarting is enabled [It] should return both restarting and running containers 
/home/aahp/go/pkg/mod/github.com/onsi/gomega@v1.21.1/ghttp/handlers.go:51

[Fail] the client when listing containers when include restarting is disabled [It] should not return restarting containers 
/home/aahp/go/pkg/mod/github.com/onsi/gomega@v1.21.1/ghttp/handlers.go:51

[Fail] the client ExecuteCommand when logging [It] should include container id field 
/home/aahp/fab/watchtower/pkg/container/client_test.go:218

Ran 44 of 44 Specs in 1.013 seconds
FAIL! -- 37 Passed | 7 Failed | 0 Pending | 0 Skipped
--- FAIL: TestContainer (1.02s)
FAIL
exit status 1
FAIL	github.com/containrrr/watchtower/pkg/container	1.028s
```

</details>

And I don't have the slightest as to why this is not working even if the container is pointing to itself.

**Comment by hicksoft** [comment]
Any updates on this PR? Updating containers with service network mode is painful right now. Looking forward to seeing this get merged.

**Comment by piksel** [comment]
@s-wasser it still needs tests, and I don't have enough knowledge of how docker network configuration changes the container info to fix it easily. I can take a look next time I get some free time though...

**Comment by shivindera** [comment]
Any updates on this getting merged?

**Comment by piksel** [comment]
@shivindera still needs tests. I have no idea if this actually works or not. The tests need to be there for us to be able to maintain this project. They tell us what some code is expected to do and how we can incorporate other changes without breaking this one's. I don't personally use this functionality myself, so it both requires extra time on reading up on and understanding the use cases, but it also means it would be guesses from my part in the end.

**Comment by heisenberg2980** [comment]
@schizo99 any chance of getting this PR tested and merged? it seems you guys were really close last year, what is missing here to fix the issue?

**Comment by schizo99** [comment]
@Heisenberg2980 What's missing are tests for the new code. But I'm getting errors when implementing the tests.
See my previous [comment](https://github.com/containrrr/watchtower/pull/1429#issuecomment-1289061655)

**Comment by heisenberg2980** [comment]
I would love to help but I have never implemented this type of tests myself. @piksel can you help with the errors @schizo99 is getting when trying to implement the test?

**Comment by piksel** [comment]
@schizo99 if you can push the tests with the errors, I can probably fix them!

Also, I got a VPN subscription and created some integration test cases that use gluetun as the container network. It still needs credentials of course, but if you have a service provider you should be able to test watchtower with it. I can add it to the tools in this PR.

**Comment by schizo99** [comment]
@piksel I never got so far as to create any test cases for these changes since it started failing as soon as I modified the `NetworkMode` for the container in `container_watchtower.json`.

So all you need to do for the tests to start failing is to update the file and set NetworkMode as this `"NetworkMode": "container:watchtower-running"`

Might very well be me not understanding the mock framework.

**Comment by piksel** [comment]
To make the tests easier to write, I did some refactoring on the docker API server mock.
I ended up just writing a single test, and the only thing that is left un-tested is when the referenced container cannot be found.
I did at least add the integration test and the corresponding `docker-compose.yml` and could confirm that it was working with this fix.
Adding an additional test for when the container isn't found should be fairly trivial, but I don't have any more time today.

**Comment by heisenberg2980** [comment]
Thank you very much @piksel, we are really close to have this PR merged and the issue fixed

**Comment by heisenberg2980** [comment]
@piksel will you be able to add the additional test to get this PR merged?

**Comment by heisenberg2980** [comment]
@piksel do you know when a new version of watchtower containing this change will be released so we can test it?

**Comment by exil0867** [comment]
> @piksel do you know when a new version of watchtower containing this change will be released so we can test it?

Use `latest-dev` image tag.

**Pull request merged**